packages/core/src/hydration/event_replay.ts

```diff
@@ -0,0 +1,95 @@
+import {unwrapRNode} from '../render3/util/view_utils';
+
+import {IS_EVENT_REPLAY_ENABLED} from './tokens';
+
+export const EVENT_REPLAY_ENABLED_DEFAULT = false;
+
+/**
+ * Returns a set of providers required to setup support for event replay.
+ * Requires hydration to be enabled separately.
+ */
+export function withEventReplay(): Provider[] {
+  return [
+    {
+      provide: IS_EVENT_REPLAY_ENABLED,
+      useValue: true,
+    },
+  ];
+}
+
+export function insertEventListenerScript(
+    appId: string, eventsToReplay: Set<string>, doc: Document) {
+  if (eventsToReplay.size) {
+    const events = Array.from(eventsToReplay);
+    const script = doc.createElement('script');
+    // This is defined in packages/core/primitives/event-dispatch/contract_binary.ts
+    script.textContent = `window.__jsaction_bootstrap('ngContracts', document.body, '${appId}', ${
+        JSON.stringify(events)});`;
+    doc.body.insertBefore(script, doc.body.firstChild);
+  }
+}
+/**
+ * Extracts information about all DOM events (added in a template) registered on elements in a give
+ * LView. Maps collected events to a corresponding DOM element (an element is used as a key).
+ */
+export function collectDomEventsInfo(
+    tView: TView, lView: LView, eventTypesToReplay: Set<string>): Map<Element, string[]> {
+  const events = new Map<Element, string[]>();
+  const lCleanup = lView[CLEANUP];
+  const tCleanup = tView.cleanup;
+  if (!tCleanup || !lCleanup) {
+    return events;
+  }
+  for (let i = 0; i < tCleanup.length;) {
+    const firstParam = tCleanup[i++];
+    const secondParam = tCleanup[i++];
+    if (typeof firstParam !== 'string') {
+      continue;
+    }
+    const name: string = firstParam;
+    eventTypesToReplay.add(name);
+    const listenerElement = unwrapRNode(lView[secondParam]) as any as Element;
+    i++;  // move the cursor to the next position (location of the listener idx)
+    const useCaptureOrIndx = tCleanup[i++];
+    // if useCaptureOrIndx is boolean then report it as is.
+    // if useCaptureOrIndx is positive number then it in unsubscribe method
+    // if useCaptureOrIndx is negative number then it is a Subscription
+    const isDomEvent = typeof useCaptureOrIndx === 'boolean' || useCaptureOrIndx >= 0;
+    if (!isDomEvent) {
+      continue;
+    }
+    if (!events.has(listenerElement)) {
+      events.set(listenerElement, [name]);
+    } else {
+      events.get(listenerElement)!.push(name);
+    }
+  }
+  return events;
+}
+
+export function setJSActionAttribute(
+    tNode: TNode, rNode: RNode, nativeElementToEvents: Map<Element, string[]>) {
+  if (tNode.type && TNodeType.Element) {
+    const nativeElement = unwrapRNode(rNode) as Element;
+    const events = nativeElementToEvents.get(nativeElement) ?? [];
+    let jsaction = '';
+    for (const event of events) {
+      jsaction += `${event}:;`;
+    }
+    if (jsaction) {
+      nativeElement.setAttribute('jsaction', jsaction);
+    }
```

Comment:
It looks like we'll be retaining an extra `;` char for each event listener, i.e. `<div jsaction="click:;">` when it should be `<div jsaction="click:">` and the `;` char is only needed when we have multiple events.

```suggestion
    const parts = events.map(event => `${event}:`);
    if (parts.length > 0) {
      nativeElement.setAttribute('jsaction', parts.join(';'));
    }
```